Enable NuGet Audit and fix issues

ericstj opened the pull request:
Microsoft.NET.HostModel can reference the live builds of the packages it depends on.  These will be deployed by the SDK.

Most other audit alerts were due to tasks pulling in old dependencies that aren't even used by the task. Avoid these by cherry-picking just the assemblies needed by the tasks and provided by MSBuild / SDK. This prevents NuGet from downloading the package closure with the vulnerable packages.  We don't need those packages since the tasks aren't responsible for deploying them.  A better solution in the future would be a targeting pack for MSBuild and the .NET SDK - so that components that contribute to these hosts have a surface area they can target without taking on responsibility for servicing.

There is once case where we have a test that references NuGet.* packages which also bring in stale dependencies that overlap with framework assemblies.  Avoid these by cherry-picking the NuGet packages in the same way.

Keeping this draft as I want to test some more and add some docs / comments - but sharing this now since a lot of folks were working on this and I wanted to share the techniques I'm using.

ericstj committed:
Enable NuGet Audit and fix issues

Microsoft.NET.HostModel can reference the live builds of the packages
it depends on.  These will be deployed by the SDK.
Most other audit alerts were due to tasks pulling in old dependencies
that aren't even used by the task. Avoid these by cherry-picking
just the assemblies needed by the tasks and provided by MSBuild / SDK.
This prevents NuGet from downloading the package closure with the
vulnerable packages.  We don't need those packages since the tasks
aren't responsible for deploying them.  A better solution in the future
would be a targeting pack for MSBuild and the .NET SDK - so that
components that contribute to these hosts have a surface area they can
target without taking on responsibility for servicing.

There is once case where we have a test that references NuGet.* packages
which also bring in stale dependencies that overlap with framework
assemblies.  Avoid these by cherry-picking the NuGet packages in the
same way.

ericstj commented on eng/PackageDownloadAndReference.targets:
```diff
@@ -0,0 +1,17 @@
+  <ItemGroup>
+    <PackageDownload Include="@(PackageDownloadAndReference)" />
+    <PackageDownload Update="@(PackageDownloadAndReference)" Version="[%(Version)]"/>
+    <Reference Include="@(PackageDownloadAndReference->'$(NuGetPackageRoot)%(Identity)/%(Version)/%(Folder)/%(AssemblyName).dll')" />
```
Looks like I need a ToLower here for linux 😆

jkoritzinsky commented on src/installer/managed/Microsoft.NET.HostModel/Microsoft.NET.HostModel.csproj:
```diff
@@ -19,8 +19,8 @@
   <ItemGroup>
-    <PackageReference Include="System.Reflection.Metadata" Version="$(SystemReflectionMetadataToolsetVersion)" />
-    <PackageReference Include="System.Text.Json" Version="$(SystemTextJsonToolsetVersion)" />
+    <ProjectReference Include="$(LibrariesProjectRoot)System.Reflection.Metadata\src\System.Reflection.Metadata.csproj" />
+    <ProjectReference Include="$(LibrariesProjectRoot)System.Text.Json\src\System.Text.Json.csproj" />
```
These references need to not move newer than MSBuild as this library is consumed by MSBuild tasks in the SDK (including for the .NET Framework build of the SDK targets). We can't reference live builds as a result.

ericstj replied:
The SDK is already on the live version of `System.Text.Json` for its tasks, but it seems it is not on the live version of `System.Reflection.Metadata`. We have another library (Microsoft.Extensions.DependencyModel) that SDK consumes with live JSON as well.  I only needed to update to live STJ to fix the audit warning so I could undo the SRM change, but it feels wrong.

I'll file a follow up issue in the SDK for that.

ViktorHofer replied:
> The SDK is already on the live version of System.Text.Json for its tasks

Would you mind elaborating? Last time I looked, NETFramework tasks in the SDK were using an 8.0.x version of STJ because MSBuild's binding redirects required that.

ericstj replied:
The SDK tasks themselves ('Microsoft.NET.Build.Tasks.dll`) reference the latest.  Because it's an exact match they don't need the redirects.  Other assemblies loaded by the SDK do reference lower versions, and they rely on MSBuild's redirects, but they don't exchange types with the SDK tasks.  The SDK also ships the latest STJ along side its tasks.
For example:
```
"C:\Program Files\dotnet\sdk\9.0.100-preview.7.24407.12\Sdks\Microsoft.NET.Sdk\tools\net472\System.Text.Json.dll" => 9.0.0.0
"C:\Program Files\dotnet\sdk\8.0.400\Sdks\Microsoft.NET.Sdk\tools\net472\System.Text.Json.dll" => 8.0.0.4
```

The SDK's MSBuild SDK resolver does need to stay on the VS copy of STJ because it doesn't carry a copy next to itself.
CC @marcpopMSFT

ericstj committed:
Fix package path on linux

ericstj committed:
Only use live JSON from HostModel

SDK pins S.R.M and a few others, so don't make them upgrade yet.

ericstj committed:
Add a couple missing assembly references

jaredpar commented on eng/PackageDownloadAndReference.targets:
```diff
@@ -0,0 +1,18 @@
+    <PackageDownload Include="@(PackageDownloadAndReference)" />
+    <PackageDownload Update="@(PackageDownloadAndReference)" Version="[%(Version)]"/>
+    <PackageDownloadAndReference Update="@(PackageDownloadAndReference)" PackageFolder="$([System.String]::new(%(Identity)).ToLowerInvariant())" />
+    <Reference Include="@(PackageDownloadAndReference->'$(NuGetPackageRoot)%(PackageFolder)/%(Version)/%(Folder)/%(AssemblyName).dll')" />
```
Won't this approach break the ability to use CPM and transitive pinning?

ericstj replied:
This doesn't pull any transitive dependencies at all.  That's the point.  It's a stand-in for a targeting pack.  We pull down just single package and use it only for reference.  It's very manual, but it avoids accidentally lifting dependencies that aren't in the control of the plugin (or pulling down dependencies just to drop them in favor of a framework library).  It should be used sparingly as it is bypassing nuget resolution.

jaredpar replied:
Sure but this means it's a different kind of package reference that you can't manage with CPM correct?

ericstj replied:
We'd just use the same properties when setting these that we'd feed into CPM.  It's a temporary solution.  We need an official package or SDK from MSBuild (and roslyn) for targeting hosts.

ericstj committed:
Refactor tasks dependencies

Consolidate representation of msbuild-provided task dependencies

ericstj committed:
Fix audit warnings in tests

ericstj commented on src/tasks/WasmAppBuilder/WasmAppBuilder.csproj:
```diff
@@ -24,27 +24,6 @@
-    <!-- we need to explicitly reference System.Reflection.MetadataLoadContext 8.0 or newer for function pointer support,
-         because the default version pulled in from MSBuild is lower. -->
-    <!-- TODO: this can be removed once MSBuild brings in a new enough version -->
-    <PackageReference Include="System.Reflection.MetadataLoadContext" Version="$(SystemReflectionMetadataLoadContextToolsetVersion)" />
```
So I tried to remove this and the build failed with 
```
Error when creating nuget packed package from D:\a\_work\1\s\artifacts\packages\Release\specs\Microsoft.NET.Runtime.WebAssembly.Wasi.Sdk.nuspec. NuGet.Packaging.Core.PackagingException: File not found: 'D:\a\_work\1\s\artifacts\bin\WasmAppBuilder\Release\net9.0\System.Reflection.MetadataLoadContext.dll'.
```

That means we were previously redistributing a non-live version of this dependency.  I think the fix would be to stop redistributing here (rather than continue redistributing the pre-built binary).

ericstj committed:
Remove MetadataLoadContext from WasmAppBuilder package

ericstj commented on eng/Versions.props:
```diff
@@ -204,7 +207,7 @@
     <GrpcDotnetClientVersion>2.45.0</GrpcDotnetClientVersion>
     <GrpcToolsVersion>2.45.0</GrpcToolsVersion>
-    <CompilerPlatformTestingVersion>1.1.2-beta1.23323.1</CompilerPlatformTestingVersion>
+    <CompilerPlatformTestingVersion>1.1.2</CompilerPlatformTestingVersion>
```
This seems to have introduced https://github.com/dotnet/roslyn-sdk/issues/1175

ericstj committed:
Update Analyzer.Testing packages

ericstj committed:
Reduce exposure of Microsoft.Build.Tasks.Core

ericstj committed:
Fix audit warnings that only occur on browser

ericstj committed:
Update Asn1 used by linker analyzer tests

ericstj committed:
Merge branch 'enableNuGetAudit' of https://github.com/ericstj/runtime into enableNuGetAudit

ericstj committed:
React to breaking change in analyzer test SDK

ericstj committed:
Enable working DryIoc tests

ericstj commented:
The build errors here look like double writes, probably my change to reference libs from hostmodel needs a fix.

ericstj committed:
Fix double-write when LibrariesConfiguration differs from Configuration

ericstj committed:
Merge branch 'main' of https://github.com/dotnet/runtime into enableNuGetAudit

ericstj committed:
Fix LibrariesConfiguration update target

ericstj committed:
Clean up references and add comments.

ericstj committed:
Make HostModel references private

This ensures projects referenced will not be rebuilt by tests.

This also means the HostModel package will not list these as references,
but that's OK since the SDK provides them and this is not a shipping
package.

ericstj committed:
Use ProjectReferenceExclusion to avoid framework project references

On .NETCore we want to use the targeting pack and avoid rebuilding libs.

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices.JavaScript/tests/JSImportGenerator.UnitTest/JSImportGenerator.Unit.Tests.csproj:
```diff
@@ -25,6 +25,9 @@
     <PackageReference Include="Microsoft.VisualBasic" Version="$(CompilerPlatformTestingMicrosoftVisualBasicVersion)" />
     <PackageReference Include="Microsoft.VisualStudio.Composition" Version="$(CompilerPlatformTestingMicrosoftVisualStudioCompositionVersion)" />
 
+    <!-- Upgrade to a non-vulnerable version of Asn1 - which will be ignored in favor of the framework copy  -->
+    <PackageReference Include="System.Formats.Asn1" Version="$(SystemFormatsAsn1Version)" />
```
Add `ExcludeAssets` like in the `ComInterfaceGenerator.Unit.Tests.csproj` project?

```suggestion
    <!-- Upgrade to a non-vulnerable version of Asn1 - which will be ignored in favor of the framework copy  -->
    <PackageReference Include="System.Formats.Asn1" Version="$(SystemFormatsAsn1Version)" ExcludeAssets="all" />
```

ericstj replied:
Sure thing.  I tried to do this to all of them, just to add confidence that we don't actually introduce a new reference or copy anywhere.

jkoritzinsky reviewed (approved):
One comment, but otherwise LGTM if everything builds.

ericstj committed:
Update src/libraries/System.Runtime.InteropServices.JavaScript/tests/JSImportGenerator.UnitTest/JSImportGenerator.Unit.Tests.csproj

Co-authored-by: Jeremy Koritzinsky <jkoritzinsky@gmail.com>
